myokit/tests/test_formats_sbml.py

```diff
@@ -15,7 +15,63 @@
 import myokit.formats.sbml
 
 # from shared import DIR_FORMATS, WarningCollector
-from myokit.tests import DIR_FORMATS, WarningCollector
+from myokit.tests import DIR_FORMATS, WarningCollector, TemporaryDirectory
+
+
+class SBMLExporterTest(unittest.TestCase):
+    """
+    Tests for :class:`myokit.formats.sbml.SBMLExporter`.
+    """
+
+    def test_capability_reporting(self):
+        # Test if the right capabilities are reported.
+        e = myokit.formats.exporter('sbml')
+        self.assertTrue(e.supports_model())
+        self.assertFalse(e.supports_runnable())
+
+    def test_stimulus_generation(self):
+        # Tests if protocols allow a stimulus current to be added
+
+        e = myokit.formats.exporter('cellml')
```

Comment:
whoops, will turn this on!

Reply:
@DavAug. The exporter uses rules to set the rhs equations for all species (no reactions). However, when I then use your parser to read in the exported file it fails at this line: https://github.com/myokit/myokit/blob/403a029a8ff05d1a847bbf14070905750c7bbafc/myokit/formats/sbml/_parser.py#L543

The SBML spec indicates that it is allowed to use rules for non boundary species, as long as you don't use reactions as well. But the code here seems to indicate that the parser will fail if rules are used for a non-boundary species?

Reply:
Hi @martinjrobins ,
I have to admit, it has been a long time since we implemented the parser. So if in doubt, I would trust you more than me! Can you specify for me what "boundary species" are? 

Just looking at the comments in the code (see below), we may have made this restriction because it's not quite obvious how you would combine assignment rules with reactions (assignments being in some some sense absolute statements about a variable's derivatives RHS, while reactions can be composed to define the RHS.

But it's very possible that we made a mistake here.

```
# Check that this assignable can be changed with assignment rules
            if isinstance(var, myokit.formats.sbml.Species):
                if not var.is_boundary():
                    raise SBMLParsingError(
                        'Assignment or rate rule set for species that is'
                        ' created or destroyed in a reaction.', element)
            # TODO: Could also gather 'constant' attributes for all, and check
            # here. But maybe we're not all that interested in validation...
```

Reply:
not really sure what boundary species are, but from the name I guess they live on the boundary and so the volume does not affect them. In any case, the SBML spec is that a boundary species can be involved in rules, or reactions, but not both (so they deal with the combination or rules + reactions by forbiding this case)

Reply:
Hi @martinjrobins and @MichaelClerx ,
I had quick refresher on the SBML specification (specifically page 51 and 52 in file:///Users/DAUGU/Downloads/sbml.level-3.version-2.core.release-2.pdf).

In particular, have a look at the last row in the table on page 52, where it says that when `constant=False` and `boundaryCondition=False`, the RHS is given by "reactions or rules (but not both), and events".

We do not support events in Myokit.

So that is why we have the above check that when we have an assignment rule to a species that is not a boundary species, we throw an error because otherwise the RHS of the species would be set by BOTH the assignment rule and the reaction.

If the species is a boundary species, the reaction does not influence the RHS of the species and the RHS is set by an assignment rule, see for example page 128 in file:///Users/DAUGU/Downloads/sbml.level-3.version-2.core.release-2.pdf

Do you agree with that understanding of the SBML documentation?

Reply:
Alternatively, (if we don't have a good way to permitting assignment rules for non boundary species) maybe it is worth trying to parse individual terms of the RHS change rates into reactions, but it would involve some more processing.

Reply:
Hey both,

I'm not sure about when a species can/should be set with an assignment rule, but I'm surprised we're talking about species at all, to be honest 😅 

The main example model we looked at for import was the HH model, which only uses "parameters": https://github.com/myokit/myokit/blob/main/myokit/tests/data/formats/sbml/HodgkinHuxley.xml
So I'd expected our output to be in terms of parameters as well.

A stronger argument is that species are defined as:

> A species in SBML refers to a pool of entities that (a) are considered indistinguishable from each other for the purposes of the model, (b) may participate in reactions, and (c) are located in a specific compartment

Section 4.6, page 49. Myokit doesn't have a concept of "reactions" or of "compartments", so whether b and c hold can't be determined from the equations alone. So the default I think, should be to export everything as a "parameter".

If we do want species to appear in the exported models, then we should think of (and document) a way to get this information into the model. For example, by doing something like:
```
dot(x) = 3
    sbml_role: species
```

Reply:
(fwiw, I think martin might be right, but have made a separate issue #1101)

Reply:
I agree. I think SBML's functionality of being able to define compartments and reactions is quite useful when you manually define large systems of ODEs, but when we automatically export myokit models, perhaps it's fine to just stick to parameters and assignment rules.

Reply:
> @martinjrobins Do you have any input/expected-output files that you're basing this on, or is it more of an "any working export will do"? Cause in that case I'd definitely ignore the complexity of species & reactions :D
> 
> (Of course, it does mean that sbml --> myokit --> sbml will make parameters out of all your species, and replace all the reactions with explicit equations. But I think that would be extremely difficult to avoid.)

All these are the inputs: https://github.com/pkpdapp-team/pkpdapp/tree/master/pkpdapp/pkpdapp/migrations/models
Outputs are just SBML files that, when simulated, give identical results the the mmt files.

I think any state variable in a myokit model should become an SBML species right? As I understand it, a paramter can't have a rate equation?

Reply:
> The use of the term parameter in SBML sometimes leads to confusion among readers who have a particular
notion of what something called “parameter” should be. It has been the source of heated debate, but despite this, no one has yet found an adequate replacement term that does not have different connotations to different people and hence leads to confusion among some subset of users. Perhaps it would have been better to have two constructs, one called “constants” and the other called “variables”. The current approach in SBML is simply more parsimonious, using a single Parameter construct with the boolean flag constant to indicate which flavor the parameter is. In any case, readers are implored to look past their particular definition of a “parameter” and simply view SBML’s Parameter as a single mechanism for defining both constants and (additional) variables in a model. (We write additional because the species in a model are usually considered to be the central variables.)

Reply:
This is Lucian's (an sbml co-author) take on it:

> Yeah, while that is technically legal, I think I would flag it as a boundary species anyway.  Or, you know, actually use 'parameter'!
>
> While the numbers in the simulation will be correct regardless, the produced model will have increased flexibility in the future if you use the correct types--when you do things like calculating the Jacobian or even just the stoichiometry matrix, having superfluous species will just confuse things.  It will also help when annotating.

So definitely parameters then

Reply:
coming back to this now. So is the plan to export all myokit variables as parameters then? What did you want me to do about compartments then, since a parameter does not have a compartment. Each parameter could just have a name based on their qname rather than their name? and otherwise we ignore compartments...?

Reply:
> So is the plan to export all myokit variables as parameters then?
Yup!

> What did you want me to do about compartments then, since a parameter does not have a compartment. 
Ignore them, I think

> Each parameter could just have a name based on their qname rather than their name? and otherwise we ignore compartments...?
Yeah it's a shame there doesn't seem to be a "conceptual" grouping mechanism, just the physical "species-compartment" one.

Reply:
although I guess this will break the expressions for species which are in compartments :)

Reply:
ahh, I guess I can just use `set_lhs_function` to set a custom name

Reply:
Sorry I was cooking :D 
Yeah that should work!